lib/puppet/util/pidlock.rb

```diff
@@ -54,7 +54,20 @@ def clear_if_stale
     begin
       Process.kill(0, lock_pid)
     rescue *errors
-      @lockfile.unlock
+      return @lockfile.unlock
+    end
+
+    # Ensure the process associated with this pid is our process. If
+    # not, we can unlock the lockfile. For now this is only done on
+    # POSIX and Windows platforms (PUP-9247).
+    if Puppet.features.posix?
+      procname = Puppet::Util::Execution.execute(["ps", "-p", lock_pid, "-o", "comm="]).strip
+      @lockfile.unlock unless procname =~ /puppet/
```

Comment:
Should we put an end of string anchor on this regex to avoid the possibility that the collision is on an infra node and puppetserver or puppetdb have grabbed the old pid?

Reply:
Good catch, I've incorporated this change and used the opportunity to squash the commits into one as well.